packages/lit-dev-content/site/articles/article/redux-reactive-controllers.md

```diff
@@ -0,0 +1,731 @@
+- [`addController(controller: ReactiveController)`](/docs/api/ReactiveElement/#ReactiveElement.addController)
+- [`removeController(controller: ReactiveController)`](/docs/api/ReactiveElement/#ReactiveElement.removeController)
+- [`requestUpdate()`](/docs/api/ReactiveElement/#ReactiveElement.requestUpdate)
+- [`updateComplete: Promise<boolean>`](/docs/api/ReactiveElement/#ReactiveElement.updateComplete)
+
+The `addController()` method takes in the controller that you want to hook into the host’s lifecycle.
+
+{% aside "info" "no-header" %}
+
+If the host is already attached to the DOM or rendered onto the page, then `hostConnected()` will be called after attaching the Reactive Controller via `addController()`.
+
+{% endaside %}
+
+A common pattern in Lit is to attach the current instance of the controller to the host in the `constructor()` of the ReactiveController. For example:
+
+```ts
+export class MyController implements ReactiveController {
+  private host: ReactiveControllerHost;
+
+  constructor(host: ReactiveControllerHost, options: MyOptions) {
+    this.host = host;
+    this.host.addController(this);
+  }
+}
+```
+
+The `removeController()` method is used less frequently than the other callbacks. It is useful when you do not want the controller to update with the host, such as: the host updates too often, the `hostUpdate[d]()` methods have slow or expensive logic, or you do not need the controller to run its updates while the component has been removed from the document.
+
+The `requestUpdate()` method is used to request the host component to re-run its update lifecycle and re-render. This is often called when the controller has a value that updates and should be reflected in the DOM. For example, the `@lit/task` package’s `Task` controller will do asynchronous work like fetching data or asynchronous rendering, and it calls the host’s `requestUpdate()` method to reflect that the state of the task has changed to pending, in progress, completed, or error which should be rendered in the component.
+
+The read-only `updateComplete` property is often used in conjunction with `requestUpdate()` method. A `ReactiveControllerHost`’s update lifecycle is assumed to be asynchronous, so the `updateComplete` property is a promise that resolves when the host’s update lifecycle has completed. This is useful for controllers that need to update the DOM and then read from it. For example, imagine a controller that resizes a DOM element and needs to then read its new dimensions. This controller would update a property, call `requestUpdate()`, await `host.updateComplete`, and then read the DOM.
+
+## Redux Patterns
+
+Redux has a bit of verbosity associated with it in order to enforce the Redux state management patterns. In this article we will be making a simple component that renders circles and squares, renders how many of them exist, and can increment or decrement the amount of circles and squares. This component will have its state managed by Redux.
```

Comment:
Commenting on this whole section: The pattern shown here is an older redux pattern that is more verbose than needed. I think the goal of this article shouldn't be to teach Redux so we just use the most concise way to create actions and reducers, which is `createSlice` from `@redux/toolkit`.
https://redux.js.org/tutorials/essentials/part-2-app-structure#the-counter-example-app

Reply:
Found the old playground example this article is based from. Here's a stab and refactoring to use `createSlice`. Auto generated action creators remove the need for the verbose enumerated action types. https://lit.dev/playground/#gist=4c07268fb5ec5fddf0280e7882d39bc0